### DISCUSSION TITLE
File upload formatter

### DISCUSSION THREAD

**Discussion by zewa666** [ghiscoding/slickgrid-universal#1726] [Ideas]
Hey there @ghiscoding 

one of the recent features is to build a workflow based grid, where certain steps are to be confirmed based on other conditions. A row represents one workflow in that case. Along with that, it will be necessary to upload files for each individual workflow (one or more).

So I was thinking to achieve this in such a way to have a string based column, which would have a custom formatter, showing a button and uploading the files. Alternatively it would open a popup with a dropzone (like seen here https://www.dropzone.dev/). Already uploaded files would be shown as tags/pills using tooltips for detail information.

Where I'm not sure actually is whether an editor is needed at all in this case. Sure one could split the "showing existing files" and showing the upload button as custom editor, but I feel like this is going to make the ui more complicated that necessary.

Additionally, I wanted to ask whether you already have used any existing file upload components you could recommend (not necessarily related to projects with the Grid), which you think could fit in nicely in a slickgrid.

**Comment by ghiscoding** [comment]
hmm I'm using Vue [Quasar](https://quasar.dev/vue-components/uploader#server-endpoint-examples) lately and barely programmed in Angular recently. The [Example 29](https://ghiscoding.github.io/slickgrid-universal/#/example29) has a drop zone using SlickGrid internal interaction classes that I wrote in plain JS in the `slick.interactions.ts` file 

https://github.com/ghiscoding/slickgrid-universal/blob/c242b388ecde51879bd990312189a5145c32c0bd/packages/common/src/core/slickInteractions.ts#L12-L29

but these interactions implementations are not related to file upload at all. It probably doesn't matter what library you use, does it? As long as you prepare the columnDefinitions/gridOptions before passing the actual data, I think you're fine to use any library you want to upload and read the data. The only other library that we use is [SortableJS](https://github.com/SortableJS/Sortable), which has drop zone too, but I'm thinking of eventually giving a try to this new lib from FormKit team [@formkit/drag-and-drop](https://drag-and-drop.formkit.com/) (I found it since they are the ones who created Tempo that I use)... but again I don't think it matters which lib you use!?

**Reply by zewa666** [reply]
yeah true, just wanted to know whether you had experiences with any of these. I guess I'// give dropzonejs a try.

what do you think about the formatter only approach I mentioned above?

**Reply by ghiscoding** [reply]
ah yeah I read it quickly and didn't notice the Formatter, not a bad idea but you should probably code your Formatter with native HTML to remain CSP compliant and avoid XSS attack on that one in particular (it's probably ok with DOMPurify as well but native code has better perf since there's nothing to parse). The tooltip for more info sounds like a good alternative to show detailed info in restrained space, I like it (and the tooltip plugin I made should handle that well with a tooltip formatter).
